File: backend/globals/classes.py
# Source: https://github.com/evilandrex/raided-loa-scraper/blob/2193cc31d8a916375029a02e9adef89e148e4196/api.py#L256
def classify_class(log: dict) -> dict:
    """Classify each player's spec based on the log"""

    def _check_skillSelfBuff(buffName: str) -> bool:
        pSelfBuffs = [
            buffCatalog[buff["buffs"][0]]["name"] for buff in pDetail["skillSelfBuffs"]
        ]
        return len([buff for buff in pSelfBuffs if buffName in buff]) > 0

    def _check_set(setName: str) -> bool:
        return f"set_{setName}" in pDetail["selfBuff"].keys()

    # Grab relevant data
    buffCatalog = log["data"]["buffCatalog"]
    skillCatalog = log["data"]["skillCatalog"]

    playerSpecs = {}
    for name, data in log["players"].items():
        # Figure out class
        pClass = data["class"]
        # Get player details
        pDetail = log["data"]["players"][name]

        if pClass == "Berserker":
            # Check if you have the Mayhem self skill buff
            playerSpecs[name] = (
                "Mayhem" if _check_skillSelfBuff("Mayhem") else "Berserker's Technique"
            )
        elif pClass == "Destroyer":
            # Look for skill "18030" special "Basic 3 Chain Hits" and does high damage split
            playerSpecs[name] = (
                "Gravity Training"
                if "18030" in pDetail["skillDamage"].keys()
                and float(pDetail["skillDamage"]["18030"]["percent"]) > 30
                else "Rage Hammer"
            )
        elif pClass == "Gunlancer":
            # First check if they're Princess Maker
            if float(log["players"][name]["percent"]) < 5:
                playerSpecs[name] = "Princess Maker"
            else:
                # Looking for set
                playerSpecs[name] = (
                    "Combat Readiness"
                    if _check_set("Hallucination") or _check_set("Nightmare")
                    else "Lone Knight"
                )
        elif pClass == "Paladin":
            # Checking if this person does okay damage
            playerSpecs[name] = (
                "Judgment"
                if float(log["players"][name]["percent"]) > 10
                else "Blessed Aura"
            )
        elif pClass == "Slayer":
            # Looking for the "Predator" skill self buff
            playerSpecs[name] = (
                "Predator" if _check_skillSelfBuff("Predator") else "Punisher"
            )
        elif pClass == "Arcanist":
            # Uses the "Emperor" skill to decide spec
            playerSpecs[name] = (
                "Order of the Emperor"
                if "19282" in pDetail["skillDamage"].keys()
                else "Empress's Grace"
            )
        elif pClass == "Summoner":
            # Check if "20290" Kelsion, is in skills
            playerSpecs[name] = (
                "Communication Overflow"
                if "20290" in pDetail["skillDamage"].keys()
                else "Master Summoner"
            )
        elif pClass == "Bard":
            # Check if they're doing okay damage
            playerSpecs[name] = (
                "True Courage"
                if float(log["players"][name]["percent"]) > 10
                else "Desperate Salvation"
            )
        elif pClass == "Sorceress":
            # Looking for "Igniter" self buff
            playerSpecs[name] = (
                "Igniter" if _check_skillSelfBuff("Igniter") else "Reflux"
            )
        elif pClass == "Wardancer":
            # Looking for esoteric skill names
            playerSpecs[name] = (
                "Esoteric Skill Enhancement"
                if any(
                    [
                        "Esoteric Skill: " in skillCatalog[id]["name"]
                        for id in pDetail["skillDamage"].keys()
                    ]
                )
                else "First Intention"
            )
        elif pClass == "Scrapper":
            # This one is weird where the Shock Training buff is ID "500224" but it doesn't have a name
            playerSpecs[name] = (
                "Shock Training"
                if len(
                    [
                        buff
                        for buff in pDetail["skillSelfBuffs"]
                        if buff["buffs"][0] == "500224"
                    ]
                )
                > 0
                else "Ultimate Skill: Taijutsu"
            )
        elif pClass == "Soulfist":
            # A weird one where the RS Hype is ID "240250" but it doesn't have a name
            playerSpecs[name] = (
                "Robust Spirit"
                if len(
                    [
                        buff
                        for buff in pDetail["skillSelfBuffs"]
                        if buff["buffs"][0] == "240250"
                    ]
                )
                > 0
                else "Energy Overflow"
            )
        elif pClass == "Glaivier":
            # Look for the "Pinnacle" skill self buff
            playerSpecs[name] = (
                "Pinnacle" if _check_skillSelfBuff("Pinnacle") else "Control"
            )
        elif pClass == "Striker":
            # Looking for the skill ID "39110", Call of the Wind God
            playerSpecs[name] = (
                "Esoteric Flurry"
                if "39110" in pDetail["skillDamage"].keys()
                else "Deathblow"
            )
        elif pClass == "Breaker":
            # Looking for the skill "47020" Asura Destruction Basic Attack
            playerSpecs[name] = (
                "Asura's Path"
                if "47020" in pDetail["skillDamage"].keys()
                else "Brawl King Storm"
            )
        elif pClass == "Deathblade":
            # Check if "25402" RE Death Trance exists and does damage
            playerSpecs[name] = (
                "Remaining Energy"
                if "25402" in pDetail["skillDamage"].keys()
                and float(pDetail["skillDamage"]["25402"]["percent"]) > 10
                else "Surge"
            )
        elif pClass == "Shadowhunter":
            # Looking for Demonic Impulse self buff
            playerSpecs[name] = (
                "Demonic Impulse"
                if _check_skillSelfBuff("Demonic Impulse")
                else "Perfect Suppression"
            )
        elif pClass == "Reaper":
            # Checking for the "Lunar Voice" self buff
            playerSpecs[name] = (
                "Lunar Voice" if _check_skillSelfBuff("Lunar Voice") else "Hunger"
            )
        elif pClass == "Souleater":
            # Checking for "Soul Snatch" self buff
            playerSpecs[name] = (
                "Night's Edge"
                if _check_skillSelfBuff("Soul Snatch")
                else "Full Moon Harvester"
            )
        elif pClass == "Sharpshooter":
            # Look for Loyal Companion skill self buff
            playerSpecs[name] = (
                "Loyal Companion"
                if _check_skillSelfBuff("Loyal Companion")
                else "Death Strike"
            )
        elif pClass == "Deadeye":
            # Look for the "Enhanced Weapon" self skill buff
            playerSpecs[name] = (
                "Enhanced Weapon"
                if _check_skillSelfBuff("Enhanced Weapon")
                else "Pistoleer"
            )
        elif pClass == "Artillerist":
            # Looking for "30260" Barrage: Focus Fire and doing more than 10% damage
            playerSpecs[name] = (
                "Barrage Enhancement"
                if "30260" in pDetail["skillDamage"].keys()
                and float(pDetail["skillDamage"]["30260"]["percent"]) > 10
                else "Firepower Enhancement"
            )
        elif pClass == "Machinist":
            # Look for Evolutionary Legacy skill self buff
            playerSpecs[name] = (
                "Evolutionary Legacy"
                if _check_skillSelfBuff("Evolutionary Legacy")
                else "Arthetinean Skill"
            )
        elif pClass == "Gunslinger":
            # Looking for Sharpshooter skill
            playerSpecs[name] = (
                "Peacemaker" if "38110" in pDetail["skillDamage"] else "Time to Hunt"
            )
        elif pClass == "Artist":
            # Checks if they're doing okay damage
            playerSpecs[name] = (
                "Recurrence"
                if float(log["players"][name]["percent"]) > 10
                else "Full Bloom"
            )
        elif pClass == "Aeromancer":
            # Check for the synergy buff "320405" sunshower on the sunshower skill "32041"
            playerSpecs[name] = (
                "Wind Fury"
                if "32041" in pDetail["skillSynergy"].keys()
                and "603_320405" in pDetail["skillSynergy"]["32041"].keys()
                else "Drizzle"
            )
        else:
            playerSpecs[name] = "Unknown"

    return playerSpecs

File: frontend/src/utils/constants/classes.ts
import { PlayerRoleType } from "@type/GeneralType";

export const SUPP_MAP: ReadonlyMap<number, string> = new Map<number, PlayerRoleType>([
    [105, "supp"],
    [204, "supp"],
    [602, "supp"],
]);

export const CLASS_ID_TO_CLASS_NAME: ReadonlyMap<number, string> = new Map<
    number,
    string
>([
    [102, "Berserker"],
    [103, "Destroyer"],
    [104, "Gunlancer"],
    [105, "Paladin"],
    [112, "Slayer"],
    [202, "Arcanist"],
    [203, "Summoner"],
    [204, "Bard"],
    [205, "Sorceress"],
    [302, "Wardancer"],
    [303, "Scrapper"],
    [304, "Soulfist"],
    [305, "Glaivier"],
    [312, "Striker"],
    [313, "Breaker"],
    [402, "Deathblade"],
    [403, "Shadowhunter"],
    [404, "Reaper"],
    [405, "Souleater"],
    [502, "Sharpshooter"],
    [503, "Deadeye"],
    [504, "Artillerist"],
    [505, "Machinist"],
    [512, "Gunslinger"],
    [602, "Artist"],
    [603, "Aeromancer"],
]);

export const CLASS_SPECS: Readonly<Record<string, ReadonlyArray<string>>> = {
    102: ["Mayhem", "Berserker's Technique"], // Berserker
    103: ["Rage Hammer", "Gravity Training"], // Destroyer
    104: ["Combat Readiness", "Lone Knight", "Princess Maker"], // Gunlancer
    105: ["Blessed Aura", "Judgment"], // Paladin
    112: ["Predator", "Punisher"], // Slayer
    202: ["Empress's Grace", "Order of the Emperor"], // Arcanist
    203: ["Master Summoner", "Communication Overflow"], // Summoner
    204: ["Desperate Salvation", "True Courage"], // Bard
    205: ["Igniter", "Reflux"], // Sorceress
    302: ["Esoteric Skill Enhancement", "First Intention"], // Wardancer
    303: ["Shock Training", "Ultimate Skill: Taijutsu"], // Scrapper
    304: ["Energy Overflow", "Robust Spirit"], // Soulfist
    305: ["Pinnacle", "Control"], // Glaivier
    312: ["Esoteric Flurry", "Deathblow"], // Striker
    313: ["Asura's Path", "Brawl King Storm"], // Breaker
    402: ["Remaining Energy", "Surge"], // Deathblade
    403: ["Demonic Impulse", "Perfect Suppression"], // Shadowhunter
    404: ["Lunar Voice", "Hunger"], // Reaper
    405: ["Night's Edge", "Full Moon Harvester"], // Souleater
    502: ["Loyal Companion", "Death Strike"], // Sharpshooter
    503: ["Enhanced Weapon", "Pistoleer"], // Deadeye
    504: ["Barrage Enhancement", "Firepower Enhancement"], // Artillerist
    505: ["Evolutionary Legacy", "Arthetinean Skill"], // Machinist
    512: ["Peacemaker", "Time to Hunt"], // Gunslinger
    602: ["Recurrence", "Full Bloom"], // Artist
    603: ["Wind Fury", "Drizzle"], // Aeromancer
};

export const CLASS_GROUPS = [
    {
        label: "Warrior",
        options: [
            { value: 102, label: "Berserker" },
            { value: 103, label: "Destroyer" },
            { value: 104, label: "Gunlancer" },
            { value: 105, label: "Paladin" },
            { value: 112, label: "Slayer" },
        ],
    },
    {
        label: "Mages",
        options: [
            { value: 202, label: "Arcanist" },
            { value: 203, label: "Summoner" },
            { value: 204, label: "Bard" },
            { value: 205, label: "Sorceress" },
        ],
    },
    {
        label: "Fighters",
        options: [
            { value: 302, label: "Wardancer" },
            { value: 303, label: "Scrapper" },
            { value: 304, label: "Soulfist" },
            { value: 305, label: "Glaivier" },
            { value: 312, label: "Striker" },
            { value: 313, label: "Breaker" },
        ],
    },
    {
        label: "Assassins",
        options: [
            { value: 402, label: "Deathblade" },
            { value: 403, label: "Shadowhunter" },
            { value: 404, label: "Reaper" },
            { value: 405, label: "Souleater" },
        ],
    },
    {
        label: "Gunners",
        options: [
            { value: 502, label: "Sharpshooter" },
            { value: 503, label: "Deadeye" },
            { value: 504, label: "Artillerist" },
            { value: 505, label: "Machinist" },
            { value: 512, label: "Gunslinger" },
        ],
    },
    {
        label: "Specialists",
        options: [
            { value: 602, label: "Artist" },
            { value: 603, label: "Aeromancer" },
        ],
    },
];

export const SUBCLASS_GROUPS = [
    {
        label: "Berserker",
        options: [
            { value: "Berserker's Technique", label: "Berserker's Technique" },
            { value: "Mayhem", label: "Mayhem" },
        ],
    },
    {
        label: "Destroyer",
        options: [
            { value: "Gravity Training", label: "Gravity Training" },
            { value: "Rage Hammer", label: "Rage Hammer" },
        ],
    },
    {
        label: "Gunlancer",
        options: [
            { value: "Combat Readiness", label: "Combat Readiness" },
            { value: "Lone Knight", label: "Lone Knight" },
            { value: "Princess Maker", label: "Princess Maker" },
        ],
    },
    {
        label: "Paladin",
        options: [
            { value: "Blessed Aura", label: "Blessed Aura -- SUPP" },
            { value: "Judgment", label: "Judgment" },
        ],
    },
    {
        label: "Slayer",
        options: [
            { value: "Predator", label: "Predator" },
            { value: "Punisher", label: "Punisher" },
        ],
    },
    {
        label: "Arcanist",
        options: [
            { value: "Empress's Grace", label: "Empress's Grace" },
            { value: "Order of the Emperor", label: "Order of the Emperor" },
        ],
    },
    {
        label: "Summoner",
        options: [
            { value: "Communication Overflow", label: "Communication Overflow" },
            { value: "Master Summoner", label: "Master Summoner" },
        ],
    },
    {
        label: "Bard",
        options: [
            { value: "Desperate Salvation", label: "Desperate Salvation -- SUPP" },
            { value: "True Courage", label: "True Courage" },
        ],
    },
    {
        label: "Sorceress",
        options: [
            { value: "Igniter", label: "Igniter" },
            { value: "Reflux", label: "Reflux" },
        ],
    },
    {
        label: "Wardancer",
        options: [
            {
                value: "Esoteric Skill Enhancement",
                label: "Esoteric Skill Enhancement",
            },
            { value: "First Intention", label: "First Intention" },
        ],
    },
    {
        label: "Scrapper",
        options: [
            { value: "Shock Training", label: "Shock Training" },
            { value: "Ultimate Skill: Taijutsu", label: "Ultimate Skill: Taijutsu" },
        ],
    },
    {
        label: "Soulfist",
        options: [
            { value: "Energy Overflow", label: "Energy Overflow" },
            { value: "Robust Spirit", label: "Robust Spirit" },
        ],
    },
    {
        label: "Glaivier",
        options: [
            { value: "Control", label: "Control" },
            { value: "Pinnacle", label: "Pinnacle" },
        ],
    },
    {
        label: "Striker",
        options: [
            { value: "Deathblow", label: "Deathblow" },
            { value: "Esoteric Flurry", label: "Esoteric Flurry" },
        ],
    },
    {
        label: "Breaker",
        options: [
            { value: "Asura's Path", label: "Asura's Path" },
            { value: "Brawl King Storm", label: "Brawl King Storm" },
        ],
    },
    {
        label: "Deathblade",
        options: [
            { value: "Remaining Energy", label: "Remaining Energy" },
            { value: "Surge", label: "Surge" },
        ],
    },
    {
        label: "Shadowhunter",
        options: [
            { value: "Demonic Impulse", label: "Demonic Impulse" },
            { value: "Perfect Suppression", label: "Perfect Suppression" },
        ],
    },
    {
        label: "Reaper",
        options: [
            { value: "Hunger", label: "Hunger" },
            { value: "Lunar Voice", label: "Lunar Voice" },
        ],
    },
    {
        label: "Souleater",
        options: [
            { value: "Full Moon Harvester", label: "Full Moon Harvester" },
            { value: "Night's Edge", label: "Night's Edge" },
        ],
    },
    {
        label: "Sharpshooter",
        options: [
            { value: "Death Strike", label: "Death Strike" },
            { value: "Loyal Companion", label: "Loyal Companion" },
        ],
    },
    {
        label: "Deadeye",
        options: [
            { value: "Enhanced Weapon", label: "Enhanced Weapon" },
            { value: "Pistoleer", label: "Pistoleer" },
        ],
    },
    {
        label: "Artillerist",
        options: [
            { value: "Barrage Enhancement", label: "Barrage Enhancement" },
            { value: "Firepower Enhancement", label: "Firepower Enhancement" },
        ],
    },
    {
        label: "Machinist",
        options: [
            { value: "Arthetinean Skill", label: "Arthetinean Skill" },
            { value: "Evolutionary Legacy", label: "Evolutionary Legacy" },
        ],
    },
    {
        label: "Gunslinger",
        options: [
            { value: "Peacemaker", label: "Peacemaker" },
            { value: "Time to Hunt", label: "Time to Hunt" },
        ],
    },
    {
        label: "Artist",
        options: [
            { value: "Full Bloom", label: "Full Bloom -- SUPP" },
            { value: "Recurrence", label: "Recurrence" },
        ],
    },
    {
        label: "Aeromancer",
        options: [
            { value: "Drizzle", label: "Drizzle" },
            { value: "Wind Fury", label: "Wind Fury" },
        ],
    },
];
